setNoMessagesReadWithResult Method

Getting an error when calling the `setNoMessagesReadWithResult` method

Package version: 2.0.7+17
Platform: Android

Error Stack:
```
D/TwilioInfo(22576): TwilioConversationsPlugin.onMethodCall => received Messages#setNoMessagesReadWithResult
D/TwilioInfo(22576): MessagesMethods.setNoMessagesReadWithResult (Channels.getChannel) => onSuccess
D/EGL_emulation(22576): app_time_stats: avg=19.39ms min=1.49ms max=195.30ms count=49
D/TwilioInfo(22576): ChannelListener.onMemberUpdated => memberSid = MB140a340eac1b42a6a9a5f9659515c714, reason = LAST_READ_TIMESTAMP
D/AndroidRuntime(22576): Shutting down VM
E/AndroidRuntime(22576): FATAL EXCEPTION: main
E/AndroidRuntime(22576): Process: com.sanadquran.sanad.dev, PID: 22576
E/AndroidRuntime(22576): com.twilio.conversations.ListenerException: Exception thrown by a listener. Your application might have a problem in listener implementation. Listeners must never throw uncaught exceptions. See 'Caused by:' below for more details.
E/AndroidRuntime(22576): 	at java.lang.reflect.Constructor.newInstance0(Native Method)
E/AndroidRuntime(22576): 	at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
E/AndroidRuntime(22576): 	at com.twilio.conversations.internal.RethrowingForwarder$RethrowingProxy.invoke(RethrowingForwarder.java:123)
E/AndroidRuntime(22576): 	at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
E/AndroidRuntime(22576): 	at $Proxy3.onSuccess(Unknown Source)
E/AndroidRuntime(22576): 	at com.twilio.conversations.internal.CallbackListenerForwarder$1.run(CallbackListenerForwarder.java:34)
E/AndroidRuntime(22576): 	at android.os.Handler.handleCallback(Handler.java:959)
E/AndroidRuntime(22576): 	at android.os.Handler.dispatchMessage(Handler.java:100)
E/AndroidRuntime(22576): 	at android.os.Looper.loopOnce(Looper.java:232)
E/AndroidRuntime(22576): 	at android.os.Looper.loop(Looper.java:317)
E/AndroidRuntime(22576): 	at android.app.ActivityThread.main(ActivityThread.java:8705)
E/AndroidRuntime(22576): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(22576): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
E/AndroidRuntime(22576): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
E/AndroidRuntime(22576): Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Number.longValue()' on a null object reference
E/AndroidRuntime(22576): 	at twilio.flutter.twilio_conversations.methods.MessagesMethods$setNoMessagesReadWithResult$1$onSuccess$1.onSuccess(MessagesMethods.kt:365)
E/AndroidRuntime(22576): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(22576): 	at com.twilio.conversations.internal.RethrowingForwarder$RethrowingProxy.invoke(RethrowingForwarder.java:121)
E/AndroidRuntime(22576): 	... 11 more
I/Process (22576): Sending signal. PID: 22576 SIG: 9
Lost connection to device.

```

--- a/flutter_twilio_conversations/CHANGELOG.md
+++ b/flutter_twilio_conversations/CHANGELOG.md
@@ -1,3 +1,17 @@
+## 2.0.9+19
+
+### Android
+* Fixed issue [#42](https://github.com/Diversido/flutter_twilio_conversations/issues/42) related to `setNoMessagesReadWithResult`
+* Fixed issue [#45](https://github.com/Diversido/flutter_twilio_conversations/issues/45) related to the namespace in Gradle builds
+* Updated Twilio Conversations Android SDK to [6.1.1](https://www.twilio.com/docs/conversations/android/changelog#conversations-611-september-2-2024)
+
+### Web
+* Fixed crash in `unregisterForNotification` if the token is empty
+* example: Updated index.html to use the latest Twilio Conversations JavaScript SDK [2.6.1](https://www.twilio.com/docs/conversations/javascript/changelog#conversations-261-february-17-2025)
+
+### General
+* Fixed issue [#44](https://github.com/Diversido/flutter_twilio_conversations/issues/44) related to `getMembersList`
+
 ## 2.0.8+18
 * Fixed issue [#41](https://github.com/Diversido/flutter_twilio_conversations/issues/41)
  

--- a/flutter_twilio_conversations/android/build.gradle
+++ b/flutter_twilio_conversations/android/build.gradle
@@ -2,7 +2,7 @@ group 'twilio.flutter.flutter_twilio_conversations'
 version '1.0-SNAPSHOT'
 
 buildscript {
-    ext.kotlin_version = '1.8.0'
+    ext.kotlin_version = '1.9.10'
     repositories {
         google()
         mavenCentral()
@@ -25,6 +25,8 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
 
 android {
+    namespace 'twilio.flutter.flutter_twilio_conversations'
+
     compileSdkVersion 33
 
     sourceSets {
@@ -45,5 +47,5 @@ android {
 dependencies {
     implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
     implementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
-    implementation 'com.twilio:conversations-android:6.0.3'
+    implementation 'com.twilio:conversations-android:6.1.1'
 }

--- a/flutter_twilio_conversations/android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/MessagesMethods.kt
+++ b/flutter_twilio_conversations/android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/MessagesMethods.kt
@@ -362,10 +362,10 @@ object MessagesMethods {
             override fun onSuccess(channel: Conversation) {
                 Log.d("TwilioInfo", "MessagesMethods.setNoMessagesReadWithResult (Channels.getChannel) => onSuccess")
 
-                channel.setAllMessagesUnread(object : CallbackListener<Long> {
-                    override fun onSuccess(index: Long) {
+                channel.setAllMessagesUnread(object : CallbackListener<Long?> {
+                    override fun onSuccess(index: Long?) {
                         Log.d("TwilioInfo", "MessagesMethods.setNoMessagesReadWithResult (Message.setNoMessagesReadWithResult) => onSuccess: $index")
-                        result.success(index)
+                        result.success(index ?: 0L)
                     }
 
                     override fun onError(errorInfo: ErrorInfo) {

--- a/flutter_twilio_conversations/example/android/build.gradle
+++ b/flutter_twilio_conversations/example/android/build.gradle
@@ -1,5 +1,5 @@
 buildscript {
-    ext.kotlin_version = '1.8.0'
+    ext.kotlin_version = '1.9.10'
     repositories {
         google()
         mavenCentral()

--- a/flutter_twilio_conversations/example/ios/Runner.xcodeproj/project.pbxproj
+++ b/flutter_twilio_conversations/example/ios/Runner.xcodeproj/project.pbxproj
@@ -254,15 +254,14 @@
 			);
 			inputPaths = (
 				"${PODS_ROOT}/Target Support Files/Pods-Runner/Pods-Runner-frameworks.sh",
-				"${BUILT_PRODUCTS_DIR}/FMDB/FMDB.framework",
 				"${BUILT_PRODUCTS_DIR}/Toast/Toast.framework",
 				"${BUILT_PRODUCTS_DIR}/device_info_plus/device_info_plus.framework",
 				"${BUILT_PRODUCTS_DIR}/flutter_local_notifications/flutter_local_notifications.framework",
 				"${BUILT_PRODUCTS_DIR}/flutter_twilio_conversations/flutter_twilio_conversations.framework",
 				"${BUILT_PRODUCTS_DIR}/fluttertoast/fluttertoast.framework",
 				"${BUILT_PRODUCTS_DIR}/image_picker_ios/image_picker_ios.framework",
-				"${BUILT_PRODUCTS_DIR}/path_provider_ios/path_provider_ios.framework",
-				"${BUILT_PRODUCTS_DIR}/sqflite/sqflite.framework",
+				"${BUILT_PRODUCTS_DIR}/path_provider_foundation/path_provider_foundation.framework",
+				"${BUILT_PRODUCTS_DIR}/sqflite_darwin/sqflite_darwin.framework",
 				"${PODS_XCFRAMEWORKS_BUILD_DIR}/TwilioCommonLib/TwilioCommonLib.framework/TwilioCommonLib",
 				"${PODS_XCFRAMEWORKS_BUILD_DIR}/TwilioConversationsClient/TwilioConversationsClient.framework/TwilioConversationsClient",
 				"${PODS_XCFRAMEWORKS_BUILD_DIR}/TwilioStateMachine/TwilioStateMachine.framework/TwilioStateMachine",
@@ -270,15 +269,14 @@
 			);
 			name = "[CP] Embed Pods Frameworks";
 			outputPaths = (
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/FMDB.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/Toast.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/device_info_plus.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/flutter_local_notifications.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/flutter_twilio_conversations.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/fluttertoast.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/image_picker_ios.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/path_provider_ios.framework",
-				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/sqflite.framework",
+				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/path_provider_foundation.framework",
+				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/sqflite_darwin.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/TwilioCommonLib.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/TwilioConversationsClient.framework",
 				"${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/TwilioStateMachine.framework",

--- a/flutter_twilio_conversations/example/ios/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
+++ b/flutter_twilio_conversations/example/ios/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
@@ -1,8 +1,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
-<plist version="1.0">
-<dict>
-	<key>IDEDidComputeMac32BitWarning</key>
-	<true/>
-</dict>
-</plist>

--- a/flutter_twilio_conversations/example/ios/Runner/AppDelegate.swift
+++ b/flutter_twilio_conversations/example/ios/Runner/AppDelegate.swift
@@ -1,7 +1,7 @@
 import UIKit
 import Flutter
 
-@UIApplicationMain
+@main
 @objc class AppDelegate: FlutterAppDelegate {
   override func application(
     _ application: UIApplication,

--- a/flutter_twilio_conversations/example/lib/redux/effects/subscriptions_middleware.dart
+++ b/flutter_twilio_conversations/example/lib/redux/effects/subscriptions_middleware.dart
@@ -38,19 +38,14 @@ class MessengerSubscriptionsMiddleware extends MiddlewareClass<AppState> {
             (event == ChatClientSynchronizationStatus.COMPLETED &&
                 store.state.chatClient?.channels != null)) {
           final dialogs = store.state.chatClient!.channels!.subscribedChannels
-              .map(
-                (channel) => ConversationDialog(
-                  channel: channel,
-                  name: channel.sid,
-                ),
-              )
-              .toList();
+              .map((channel) {
+            return ConversationDialog(
+              channel: channel,
+              name: channel.sid,
+            );
+          }).toList();
 
-          store.dispatch(
-            UpdateDialogsAction(
-              dialogs,
-            ),
-          );
+          store.dispatch(UpdateDialogsAction(dialogs));
 
           for (var conversation
               in store.state.chatClient!.channels!.subscribedChannels) {
@@ -58,6 +53,10 @@ class MessengerSubscriptionsMiddleware extends MiddlewareClass<AppState> {
             store.dispatch(
                 GetConversationUnreadMessagesCountAction(conversation));
             store.dispatch(SubscribeToMembersTypingStatus(conversation));
+
+            //Only for example purposes, print members
+            final members = await conversation.members?.getMembersList();
+            debugPrint('Got members: ${members?.length}');
           }
 
           store.dispatch(

--- a/flutter_twilio_conversations/example/lib/redux/effects/subscriptions_middleware.dart
+++ b/flutter_twilio_conversations/example/lib/redux/effects/subscriptions_middleware.dart
@@ -38,12 +38,19 @@ class MessengerSubscriptionsMiddleware extends MiddlewareClass<AppState> {
             (event == ChatClientSynchronizationStatus.COMPLETED &&
                 store.state.chatClient?.channels != null)) {
           final dialogs = store.state.chatClient!.channels!.subscribedChannels
-              .map((channel) {
-            return ConversationDialog(
-              channel: channel,
-              name: channel.sid,
-            );
-          }).toList();
+              .map(
+                (channel) => ConversationDialog(
+                  channel: channel,
+                  name: channel.sid,
+                ),
+              )
+              .toList();
+
+          store.dispatch(
+            UpdateDialogsAction(
+              dialogs,
+            ),
+          );
 
           store.dispatch(UpdateDialogsAction(dialogs));
 

--- a/flutter_twilio_conversations/example/lib/redux/effects/subscriptions_middleware.dart
+++ b/flutter_twilio_conversations/example/lib/redux/effects/subscriptions_middleware.dart
@@ -52,8 +52,6 @@ class MessengerSubscriptionsMiddleware extends MiddlewareClass<AppState> {
             ),
           );
 
-          store.dispatch(UpdateDialogsAction(dialogs));
-
           for (var conversation
               in store.state.chatClient!.channels!.subscribedChannels) {
             store.dispatch(GetConversationMessagesAction(conversation));

--- a/flutter_twilio_conversations/example/web/index.html
+++ b/flutter_twilio_conversations/example/web/index.html
@@ -37,8 +37,8 @@
     // The value below is injected by flutter build, do not touch.
     const serviceWorkerVersion = null;
   </script>
-  <script src="https://sdk.twilio.com/js/conversations/releases/2.5.0/twilio-conversations.min.js"
-    integrity="sha256-aBKm/Hjylgtmr/XtFMSDbUWf+2VlRHBbj4jdgy/KhRc=" crossorigin="anonymous"></script>
+  <script src="https://sdk.twilio.com/js/conversations/releases/2.6.1/twilio-conversations.min.js"
+   integrity="sha256-dvIsyBhuA90h9qTVN37CZ1kLwp+JZRqeV/lxv8X0SoA="crossorigin="anonymous"></script>
   <!-- This script adds the flutter initialization JS code -->
   <script src="flutter.js" defer></script>
 </head>

--- a/flutter_twilio_conversations/ios/.gitignore
+++ b/flutter_twilio_conversations/ios/.gitignore
@@ -35,3 +35,5 @@ Icon?
 
 /Flutter/Generated.xcconfig
 /Flutter/flutter_export_environment.sh
+
+.build

--- a/flutter_twilio_conversations/lib/src/attributes.dart
+++ b/flutter_twilio_conversations/lib/src/attributes.dart
@@ -12,11 +12,11 @@ class Attributes {
 
   Attributes(this._type, this._json);
 
-  factory Attributes.fromMap(Map<String, dynamic> map) {
-    final type =
-        EnumToString.fromString(AttributesType.values, map['type'] ?? 'null') ??
-            AttributesType.NULL;
-    return Attributes(type, map['data'] ?? '');
+  factory Attributes.fromMap(Map<String, dynamic>? map) {
+    final type = EnumToString.fromString(
+            AttributesType.values, map?['type'] ?? 'null') ??
+        AttributesType.NULL;
+    return Attributes(type, map?['data'] ?? '');
   }
 
   Map<String, dynamic>? getJSONObject() {

--- a/flutter_twilio_conversations/lib/src/member.dart
+++ b/flutter_twilio_conversations/lib/src/member.dart
@@ -68,7 +68,7 @@ class Member {
       map['sid'],
       EnumToString.fromString(MemberType.values, map['type'])!,
       map['channelSid'],
-      Attributes.fromMap(map['attributes'].cast<String, dynamic>()),
+      Attributes.fromMap(map['attributes']?.cast<String, dynamic>()),
     );
     member._updateFromMap(map);
     return member;

--- a/flutter_twilio_conversations/lib/src/members.dart
+++ b/flutter_twilio_conversations/lib/src/members.dart
@@ -19,7 +19,7 @@ class Members {
   Future<List<Member>?> getMembersList() async {
     final membersListData = await FlutterTwilioConversationsPlatform.instance
         .getMembersList(_channelSid);
-    if (membersListData['membersList'] != null) {
+    if (membersListData?['membersList'] != null) {
       return membersListData['membersList']
           .map<Member>((m) => Member._fromMap(m.cast<String, dynamic>()))
           .toList();

--- a/flutter_twilio_conversations/pubspec.yaml
+++ b/flutter_twilio_conversations/pubspec.yaml
@@ -1,6 +1,6 @@
 name: flutter_twilio_conversations
 description: Integrate the Twilio Conversations SDK with your Flutter app using this Twilio Conversations Flutter plugin
-version: 2.0.8+18
+version: 2.0.9+19
 repository: https://github.com/Diversido/flutter_twilio_conversations
 issue_tracker: https://github.com/Diversido/flutter_twilio_conversations/issues
 
@@ -11,8 +11,8 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
-  flutter_twilio_conversations_platform_interface: ^2.0.8
-  flutter_twilio_conversations_web: ^2.0.8
+  flutter_twilio_conversations_platform_interface: ^2.0.9
+  flutter_twilio_conversations_web: ^2.0.9
   enum_to_string: ^2.0.1
   intl: ^0.19.0
 

--- a/flutter_twilio_conversations/test/member_test.dart
+++ b/flutter_twilio_conversations/test/member_test.dart
@@ -1,0 +1,55 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:flutter_twilio_conversations/flutter_twilio_conversations.dart';
+
+void main() {
+  setUpAll(() {});
+
+  group('.fromMap()', () {
+    test('should return null when map is null', () {
+      final member = Member.fromMap(null);
+      expect(member, isNull);
+    });
+
+    test('should create a valid Member from a complete map', () {
+      final input = {
+        'sid': 'SM123',
+        'type': 'chat',
+        'channelSid': 'CH123',
+        'attributes': {'key': 'value'},
+        'lastReadMessageIndex': 5,
+        'lastConsumptionTimestamp': '2023-10-05T12:00:00Z',
+        'identity': 'user1',
+      };
+
+      final member = Member.fromMap(input);
+      expect(member, isNotNull);
+      expect(member!.sid, equals('SM123'));
+      expect(member.type.toString().toLowerCase(), contains('chat'));
+      expect(member.attributes, isNotNull);
+      expect(member.lastReadMessageIndex, equals(5));
+      expect(member.lastConsumptionTimestamp, equals('2023-10-05T12:00:00Z'));
+      expect(member.identity, equals('user1'));
+    });
+
+    test('should create a valid Member from with no attributes', () {
+      final input = {
+        'sid': 'SM123',
+        'type': 'chat',
+        'channelSid': 'CH123',
+        'lastReadMessageIndex': 5,
+        'lastConsumptionTimestamp': '2023-10-05T12:00:00Z',
+        'identity': 'user1',
+      };
+
+      final member = Member.fromMap(input);
+      expect(member, isNotNull);
+      expect(member!.sid, equals('SM123'));
+      expect(member.type.toString().toLowerCase(), contains('chat'));
+      expect(member.attributes!.type, AttributesType.NULL);
+      expect(member.attributes!.getJSONArray(), null);
+      expect(member.lastReadMessageIndex, equals(5));
+      expect(member.lastConsumptionTimestamp, equals('2023-10-05T12:00:00Z'));
+      expect(member.identity, equals('user1'));
+    });
+  });
+}

--- a/flutter_twilio_conversations/test/member_test.dart
+++ b/flutter_twilio_conversations/test/member_test.dart
@@ -31,7 +31,7 @@ void main() {
       expect(member.identity, equals('user1'));
     });
 
-    test('should create a valid Member from with no attributes', () {
+    test('should create a valid Member from a map with no attributes', () {
       final input = {
         'sid': 'SM123',
         'type': 'chat',

--- a/flutter_twilio_conversations_platform_interface/pubspec.yaml
+++ b/flutter_twilio_conversations_platform_interface/pubspec.yaml
@@ -1,6 +1,6 @@
 name: flutter_twilio_conversations_platform_interface
 description: A common platform interface for the flutter_twilio_conversations plugin.
-version: 2.0.8
+version: 2.0.9
 repository: https://github.com/Diversido/flutter_twilio_conversations/tree/main/flutter_twilio_conversations_platform_interface
 issue_tracker: https://github.com/Diversido/flutter_twilio_conversations/issues
 

--- a/flutter_twilio_conversations_web/CHANGELOG.md
+++ b/flutter_twilio_conversations_web/CHANGELOG.md
@@ -1,3 +1,7 @@
+## 2.0.9
+* Fix crash with `unregisterForNotification` if the token is empty
+* Update example to use the latest Twilio Conversations JavaScript SDK [2.6.1](https://sdk.twilio.com/js/conversations/releases/2.6.1/docs/index.html)
+
 ## 2.0.6
 * Fix crash with `typing`, `getUnreadMessagesCount`, `connectionStateChanged` and `connectionError` methods
 * Refactored `sendMessage` to improve stability during conversation creation

--- a/flutter_twilio_conversations_web/lib/methods/notifications_methods.dart
+++ b/flutter_twilio_conversations_web/lib/methods/notifications_methods.dart
@@ -1,4 +1,5 @@
 import 'package:flutter_twilio_conversations_web/interop/classes/client.dart';
+import 'package:flutter_twilio_conversations_web/logging.dart';
 import 'package:js/js_util.dart';
 
 class NotificationsMethods {
@@ -17,6 +18,11 @@ class NotificationsMethods {
     required String channel,
     required String token,
   }) async {
+    if (token.isEmpty) {
+      Logging.debug('error: the parameter "token" was not given');
+      return;
+    }
+
     await promiseToFuture<void>(
       chatClient?.removePushRegistrations(channel, token),
     );

--- a/flutter_twilio_conversations_web/pubspec.yaml
+++ b/flutter_twilio_conversations_web/pubspec.yaml
@@ -1,6 +1,6 @@
 name: flutter_twilio_conversations_web
 description: Web platform implementation of flutter_twilio_conversations_web
-version: 2.0.8
+version: 2.0.9
 repository: https://github.com/Diversido/flutter_twilio_conversations/tree/main/flutter_twilio_conversations_web
 issue_tracker: https://github.com/Diversido/flutter_twilio_conversations/issues
 
@@ -17,7 +17,7 @@ dependencies:
     sdk: flutter
   flutter_web_plugins:
     sdk: flutter
-  flutter_twilio_conversations_platform_interface: ^2.0.8
+  flutter_twilio_conversations_platform_interface: ^2.0.9
   meta: ^1.12.0
   js: ^0.6.4
   intl: ^0.19.0

